### DISCUSSION TITLE
Bundle Explorer (prototype) shows modules, sizes from InstallBundle tx

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#8416] [Show and tell]
@anilhelvaci asked about [the cost of deploying bundles](https://github.com/Agoric/agoric-sdk/discussions/8285) and if there's a way to reduce bundle size. I offered to take a look at their bundle for opportunities.

I made a quick-n-dirty bundle explorer:
 - experimental deploy: https://bexp.netlify.app/
 - https://github.com/agoric-labs/bundle-tool2

It started as a [bundle explorer gist](https://gist.github.com/dckc/5eb31c216f4bf10e8f7a0ff82ced6108) to look at [the swingset.MsgInstallBundle transaction for the crabble contract](https://devnet.explorer.agoric.net/agoric/tx/97B09AD76E94D9FACC9FCF21D97364195B4E746DC76307697CBF70C35378FB07) on `agoricdev-21`.

![screenshot of bundle explorer showing modules and their sizes](https://github.com/Agoric/agoric-sdk/assets/150986/84bf33b4-ef60-4b98-bc34-aea8ef225fd5).

To use it, you currently have to download [unbundle.html](https://gist.github.com/dckc/5eb31c216f4bf10e8f7a0ff82ced6108#file-unbundle-html) an [unbundle.js](https://gist.github.com/dckc/5eb31c216f4bf10e8f7a0ff82ced6108#file-unbundle-js) and use a static HTTP server. I suppose I could host it somewhere...

Browsing through the bundle contents, I'm reminded that in due course, you should only need to pay for new modules:

 - https://github.com/Agoric/agoric-sdk/issues/2391

In the shorter term, there's some possible clean-up we are looking into:

 - https://github.com/Agoric/agoric-sdk/issues/6512

cc @kriskowal @turadg @warner

**Comment by dckc** [comment]
deployed at https://bundle-explorer1.netlify.app/unbundle.html

also:
 - handles other nodes (e.g. main.rpc.agoric.net)
 - computes bundle storage cost

For example: KREAd aux proposal bundle tx [95B037...](https://agoric.explorers.guru/transaction/95B037ED064FC5A3FE95AD535BABE4A59A263868042614BCCC00F1AAE5C3AD96)

![image](https://github.com/Agoric/agoric-sdk/assets/150986/a19bd37c-7124-4ce6-9fc2-f6afaac5a0b3)

**Comment by dckc** [comment]
I looked into [differential bundle loading](https://gist.github.com/dckc/5b2e9faf0cb6028e09bec0d12eeea383).

The last half of the rows in [summary.csv](https://gist.github.com/dckc/5b2e9faf0cb6028e09bec0d12eeea383#file-summary-csv) suggest that we'd save about 90% on contract uploads if we exclude stuff that's already on chain.

And the remaining 10% is dominated by `compartment-map.json`.

**Comment by toliaqat** [comment]
@dckc thanks for Bundle Explorer. I took the code and ported it to the explorer I am playing with using Appsmith.

Following features were added in it.
* Load transactions that have bundle installed in them
* Capability to inspect the content of the code files in the bundle
* Compare the bundle code with github version
* Share the bundle links with others

https://swings.live/app/bundles-65206cff5a89f25b286288e0

![Screenshot 2023-10-10 at 12 12 57 PM](https://github.com/Agoric/agoric-sdk/assets/6778940/e60e5b22-1572-419a-ac95-afba7166db93)

**Comment by dckc** [comment]
I added showing the bundle hash in the gist
haven't re-deployed to netlify

**Comment by dckc** [comment]
I'm starting to add bundle upload. Things were getting a little complex, so I started migrating to vite + lit-ts

https://github.com/agoric-labs/bundle-tool2

 - 2023-10-22 09:47 7f4c203 chore: use bundle elements
 - ...
 - 2023-10-01 01:43 4003ddd

**Comment by dckc** [comment]
I tried adding a [download as zip feature](https://github.com/agoric-labs/bundle-tool2/issues/1), but got bogged down.

_IOU a copy of my WIP._
